### PR TITLE
GH-15164: [C++][Parquet] Implement current version of BloomFilter spec

### PR DESCRIPTION
**Pull request by mapleFU**
The original Parquet Bloom Filter spec, added in 2018 (PARQUET-41), was based on the Murmur3 hash function. The spec was later changed to use the XXH64 hash function from xxHash (PARQUET-1609); however, Parquet C++ wasn't updated for this and kept implementing the original spec.

This PR switches the bloom filter implementation to the current version of the Bloom Filter spec (as of https://github.com/apache/parquet-format/commit/3fb10e00c2204bf1c6cc91e094c59e84cefcee33). Conformance is tested using a dedicated test file in the parquet-testing repository (`bloom_filter.xxhash.bin`).

### CODE REVIEW

**Comment by mapleFU** [comment]
@pitrou @wjones127 @emkornfield Mind take a look?

**Comment by mapleFU** [comment]
Well, why should I have a assignee but I cannot edit the description of patch? Should someone be assignee or assign this task to me? ╮(￣▽￣"")╭

**Comment by mapleFU** [comment]
I generate a test file, and update it here: https://github.com/apache/parquet-testing/pull/34

You can merge that. And after that, we can run testing here.

**Comment by rok** [comment]
> Well, why should I have a assignee but I cannot edit the description of patch? Should someone be assignee or assign this task to me? ╮(￣▽￣"")╭

I've assigned it to you. Can you check if you can edit it now?

File: cpp/src/parquet/bloom_filter.h
Comment (wjones127) [inline cpp/src/parquet/bloom_filter.h:124]: Should this be updated?
```suggestion
  /// The constructor of BlockSplitBloomFilter. It uses XXH64 as hash function.
```

**Review by wjones127 — COMMENTED** [review]
One suggestion on a doc comment but otherwise looks good. I don't have merge permission on the testing repo, so someone else will need to merge that to unblock you. @pitrou ?

**Comment by mapleFU** [comment]
Thanks wjones! Comment fixed. Waiting for test file to be merged.

**Comment by mapleFU** [comment]
> I've assigned it to you. Can you check if you can edit it now?

Thansk @rok. Yes I can unassign now. Let me try it in coming patches later

**Comment by mapleFU** [comment]
```
error: failed to commit transaction (conflicting files)
mingw-w64-ucrt-x86_64-openssl: /ucrt64/bin/libcrypto-3-x64.dll exists in filesystem
mingw-w64-ucrt-x86_64-openssl: /ucrt64/bin/libssl-3-x64.dll exists in filesystem
mingw-w64-ucrt-x86_64-openssl: /ucrt64/etc/ssl/openssl.cnf exists in filesystem
Errors occurred, no packages were upgraded.
Error: Process completed with exit code 1.
```

Seems windows failed is not cause by me?

File: cpp/src/parquet/xxhash.h
Comment (wgtmac) [inline cpp/src/parquet/xxhash.h:29]: Do we need this comment at all? xxhash is already vendored in this repo.

File: cpp/src/parquet/xxhash.h
Comment (wgtmac) [inline cpp/src/parquet/xxhash.h:31]:
```suggestion
class PARQUET_EXPORT XxHasher : public Hasher {
```

File: cpp/src/parquet/xxhash.h
Comment (wgtmac) [inline cpp/src/parquet/xxhash.h:40]: Is it more convenient to add overload `uint64_t Hash(std::string_view value) const = 0` to the base class? Then user does not need to convert to `ByteArray` or `FLBA` if `string_view` is on hand.

File: cpp/src/parquet/murmur3.cc
Comment (wgtmac) [inline cpp/src/parquet/murmur3.cc:190]: Can we delete `mormur3.cc`? I assume it is not used anywhere.

File: cpp/src/parquet/xxhash.h
Comment (mapleFU) [reply cpp/src/parquet/xxhash.h:29]: Just as a testing. These code are just equal-to xxhash bloom testing in parquet-mr.

File: cpp/src/parquet/xxhash.h
Comment (mapleFU) [reply cpp/src/parquet/xxhash.h:40]: Maybe we can hash them using string_view internally?

Comment (wgtmac) [reply cpp/src/parquet/xxhash.h:40]: That works. My point is about adding a public function so that user does not have to convert if the data is already typed std::string_view. Current functions are more likely to be used internally in parquet.

Comment (mapleFU) [reply cpp/src/parquet/xxhash.h]: I think we can add it later if any one needs it, because we need to add `Hash(std::string_view)` in hasher and it subclass, and if anyone need it, he/she can add it

**Comment by mapleFU** [comment]
@pitrou @wjones127 @wgtmac The parquet-testing is updated, and comment resolved, PTAL?

**Comment by mapleFU** [comment]
(This request for review is so interesting...when I invite a mentor, another one would be revoked...)

**Review by wjones127 — APPROVED** [review]
This looks good to me! I'll give @wgtmac a chance to give a final review before merging.

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:78]:
```suggestion
static ::arrow::Status ValidateBloomFilterHeader(
```

Comment (wgtmac) [reply cpp/src/parquet/bloom_filter.cc:78]: Usually we use capitalized initials.

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:75]:
```suggestion
static constexpr uint32_t kHeaderSizeGuess = 32;
```

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:76]:
```suggestion
static constexpr uint32_t kMaxBloomFilterHeaderSize = 1024;
```

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:112]: Remove blank line

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:110]: This should be an input parameter.

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:113]:
```suggestion
  uint32_t header_size = 0;
```

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:162]:
```suggestion
    throw ParquetException("BloomFilter does not support Hash other than XXHASH");
```

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:166]:
```suggestion
    throw ParquetException("BloomFilter does not support Compression other than UNCOMPRESSED");
```

File: cpp/src/parquet/parquet.thrift
Comment (wgtmac) [inline cpp/src/parquet/parquet.thrift:588]: Should we update the generated cop as well? And is it good time to sync the latest thrift file from upstream parquet-format? @wjones127 @pitrou

File: cpp/src/parquet/xxhasher.h
Comment (wgtmac) [inline cpp/src/parquet/xxhasher.h:41]:
```suggestion
  static constexpr int kDefaultSeed = 0;
```

**Review by wgtmac — COMMENTED** [review]
It looks good. I have left some minor cosmetic comments. Thanks @mapleFU for working on it and @wjones127 for mentioning me!

File: cpp/src/parquet/parquet.thrift
Comment (wjones127) [reply cpp/src/parquet/parquet.thrift:588]: Actually is this change necessary?

We could sync the file. It just has updated comments with clarifications and some deprecations.

Comment (wgtmac) [reply cpp/src/parquet/parquet.thrift:588]: Double checked the upstream. This line should be reverted.

**Comment by mapleFU** [comment]
@wgtmac Comment updated, mind take a look?

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:110]:
```suggestion
    const ReaderProperties& properties, ArrowInputStream* input) {
```

File: cpp/src/parquet/bloom_filter.cc
Comment (wgtmac) [inline cpp/src/parquet/bloom_filter.cc:147]:
```suggestion
  BlockSplitBloomFilter bloom_filter;
```

**Review by wgtmac — COMMENTED** [review]
Some minor comments. Others look good to me. Thanks!

**Comment by mapleFU** [comment]
Change finished, some test failed because S3...

**Review by wgtmac — APPROVED** [review]
Thanks @mapleFU! 
LGTM

**Comment by mapleFU** [comment]
Let me retrigger it once, by the way, it's there any command, that I can rerun a unittest? Like 

`/rerun C++ / AMD64 Windows MinGW MINGW32 C++ (pull_request) `

Seems that failed in S3 will making all test to be rerun, which would be so anonying.

**Comment by mapleFU** [comment]
add description in `BloomFIlter::Deserialize`, because it will make doctest failed if no argument provided...

**Comment by mapleFU** [comment]
```
9/90 Test #20: arrow-io-buffered-test ....................   Passed    0.53 sec
      Start 21: arrow-io-compressed-test
Error: The operation was canceled.
```

and arrow-array-test:

```
[----------] Global test environment tear-down
[==========] 834 tests from 95 test suites ran. (53645 ms total)
[  PASSED  ] 834 tests.

  YOU HAVE 2 DISABLED TESTS

Found core dump, printing backtrace:
(lldb) target create --core "core.arrow-array-test.41874"
```

let me figure out why...

**Comment by mapleFU** [comment]
@wjones127 I've finish the comment fix, but test still failed some times, and I think they're unrelated to my patch...

**Comment by mapleFU** [comment]
@wjones127 All tests green, can we merge this patch now?

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:119]: We should use `Read` as `Peek` is not guaranteed to be implemented by all input streams.

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:76]: I think this is pointless complication. We can try to read up to 256 bytes at once, which should be sufficient for future additions.

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:89]: You should be able to write this more easily:
```suggestion
    return ::arrow::Status::Invalid("Unsupported Bloom filter hash: ", header.hash, ".";);
```

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:141]: Why not `PARQUET_THROW_NOT_OK`?

File: cpp/src/parquet/bloom_filter.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter.cc:89]: Yeah, this is add by myself but I forgot it, lol...

File: cpp/src/parquet/bloom_filter.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter.cc:76]: I think 32B should be enough for a long time...

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:145]: Should also check that buffer size is equal to `header.numBytes`.

File: cpp/src/parquet/bloom_filter.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter.cc:119]: I think `SerializedPageReader` already uses `Peek`, so at least in parquet, using `Peek` is ok

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:195]: This would be better as:
```suggestion
  const uint32_t* bitset32 = reinterpret_cast<const uint32_t*>(data_->data());
```

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:194]: Make this `const`?

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:212]: `const` here too?

File: cpp/src/parquet/bloom_filter.h
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.h:204]: Should reorder these two lines

File: cpp/src/parquet/bloom_filter.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter.cc:212]: Seems not, `Insert` may modify bitset...

File: cpp/src/parquet/bloom_filter_test.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter_test.cc:168]: Why remove this line? You should instead update it to correctly explain the length value.

File: cpp/src/parquet/bloom_filter_test.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter_test.cc:239]: @pitrou these test is pasted from parquet-mr: https://github.com/apache/parquet-mr/commit/806037c080dc477798d157cd4a54a81240a85d37#diff-49a81f9fa74dcf80bbb13202cf9bf4c94ace2030920a2fce326b4ba879749c7e

File: cpp/src/parquet/xxhasher.h
Comment (pitrou) [inline cpp/src/parquet/xxhasher.h:30]: Just for the record, is the seed parameter actually useful? The Parquet spec mandates seed value 0. Hardcoding the seed might make hashing slightly faster.

Comment (mapleFU) [reply cpp/src/parquet/xxhasher.h:30]: Seems that release will optimize this. I just keep like previous MurmurHash3, since `xxhasher` and it's base class "hasher" is not `parquet_xxhasher`.

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [reply cpp/src/parquet/bloom_filter.cc:76]: What I mean is that more generally we can read a fixed size and simply error out if deserializing fails.

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [reply cpp/src/parquet/bloom_filter.cc:115]: I'm not sure, because the InputStream might have been wrapped in a BufferedReader internally.

File: cpp/src/parquet/bloom_filter_test.cc
Comment (pitrou) [reply cpp/src/parquet/bloom_filter_test.cc:240]: Well, we don't need to copy blindly what parquet-mr does :-)

File: cpp/src/parquet/xxhasher.h
Comment (mapleFU) [reply cpp/src/parquet/xxhasher.h:30]: https://godbolt.org/z/bjz5WGx83 Seems maybe it will. Still I think `Hasher` is not ParquetHasher, the name of it is too

File: cpp/src/parquet/bloom_filter_test.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter_test.cc:240]: > Well, we don't need to copy blindly what parquet-mr does :-)

I think it's correct and standard, and it will be too long, just copying this piece of code is much trustworthy than hello world and different seed...

File: cpp/src/parquet/xxhasher.h
Comment (pitrou) [reply cpp/src/parquet/xxhasher.h:30]: > https://godbolt.org/z/bjz5WGx83 Seems maybe it will.

I don't think this example is proving anything as 1) everything is implemented in the  `.h` (thus trivially inlinable) 2) the result value doesn't even depend on the seed.

> Still I think `Hasher` is not ParquetHasher, the name of it is too

No, but it's in the `src/parquet` directory, which is a good hint :-)

File: cpp/src/parquet/bloom_filter.cc
Comment (mapleFU) [reply cpp/src/parquet/bloom_filter.cc:76]: Emmmm, I got what you mean...I want use `Read` instead of `Peek` for compability, and can only `Read` once to make the program simplier...

Comment (pitrou) [reply cpp/src/parquet/bloom_filter.cc:76]: Yes, exactly ;-)

**Comment by mapleFU** [comment]
@pitrou I implement it using `Read` rather than `Seek`, to be honest, it make the program much more complex, you can review and modify it.

File: cpp/src/parquet/bloom_filter.cc
Comment (pitrou) [inline cpp/src/parquet/bloom_filter.cc:132]: These could just be equal. So I would say `DCHECK_GE(sv->size(), header_size)`.

File: cpp/src/parquet/bloom_filter.cc
Comment (mapleFU) [inline cpp/src/parquet/bloom_filter.cc:141]: If `header_size == header_buf->size()`, here would call `memcpy(len = 0)`. `len = 0` in memcpy is valid if `src` and `data` has valid value.

So I'm afraid that it will call memcpy with 0 on a dangerous address... @pitrou

Comment (pitrou) [reply cpp/src/parquet/bloom_filter.cc:141]: That's a good point. I will revert this one.

**Pull request merged**
